v2/clients/http/utils/request.go

```diff
@@ -31,13 +33,33 @@ func GetRequest(ctx context.Context, returnValuePointer interface{}, baseUrl str
 	return nil
 }
 
+func setApiVersion(data interface{}) {
```

Comment:
Why do we need this when we have added the New factory method. I assume there will be new factory methods for the other Requests for consistency?

Reply:
Our purpose is to prevent sending POST, PUT, PATCH request without API version.
- **NewAddEventRequest**: Services need to use NewAddEventRequest to create a AddEventRequest and send the request via message bus.
- **setApiVersion**: Otherwise we can send request via client library and the setApiVersion func can help to add the api version.

Reply:
I would rather see ApiVersion as a required field rather than doing reflection to make sure it is set. i.e. if this is the need, why not use validation like on other fields for make sure it is set properly.

@cloudxxx8 , thoughts?

Reply:
@lenny-intel I agree with setting ApiVersion as a required field, but we can still make sure everything using the client library would set the ApiVersion automatically.  If the users use V2 client library, they are using V2 API, not other version.

I changed my mind about creating new factory methods for request, because there will be lots of input parameters, much more than response.  It would be hard to use.

Reply:
> I changed my mind about creating new factory methods for request, because there will be lots of input parameters, much more than response. It would be hard to use.

@cloudxxx8 , yea I was thinking that would be an issue, but thought that the factory methods could be simple, i.e. no parameters and set ApiVersion, requestID, etc, and then dev has to set the rest of the properties as is done now. Thoughts?

Reply:
> I agree with setting ApiVersion as a required field,

@cloudxxx8 , if ApiVersion is required, then it should be set already otherwise it fails validation. So no need to use reflection to make sure it is set.

Reply:
@lenny-intel the client library here is the caller, so it won't validate before sending out the request.  This func can help make sure that every request send out from the client library with the correct api version.

> 
> @cloudxxx8 , yea I was thinking that would be an issue, but thought that the factory methods could be simple, i.e. no parameters and set ApiVersion(s), IDs, etc, and then dev has to set the rest of the properties as is done now. Thoughts?

Could you please write a simple example?

Reply:
@cloudxxx8 , @weichou1229 , I have created factory methods for `BaseRequest`, `AddEventRequest`, `Event` and `Reading` in this branch: 
https://github.com/edgexfoundry/go-mod-core-contracts/compare/master...lenny-intel:event-reading-factories

This example does have parameters for most properties and shows how some properties, such as `ApiVersion`,  can be auto filled and others left to be filled in later, such as `Readings`.
I still need to add some unit tests.
If you like this approach I will finish the unit tests and submit a PR.

For the other request factory methods, `Name` should be a parameter and the rest can be left to fill in later. Just depends on the request and its sub-dtos on what works best.

Reply:
@cloudxxx8 , @weichou1229 , I have created a PR for my added factory methods for AddEventRequest/Event/Reading DTOs.
https://github.com/edgexfoundry/go-mod-core-contracts/pull/468

Note that I think the factory methods for the other Request and sub DTOs  may be simpler and leave setting bulk of the properties after New() has returned the DTO.

Reply:
> func NewDeviceRequest (deviceName string, profileName string, serviceName string, adminState string, operatingState string)  AddDeviceRequest 

I thought the method NewDeviceRequest is too many parameters. Can we use few parameters like NewDeviceResponse?
```
func NewAddDeviceRequest(device dtos.Device)(AddDeviceRequest,errors.EdgeX){
	req := AddDeviceRequest{
		BaseRequest: common.NewBaseRequest(),
		Device:       device,
	}

	err := req.Validate()
	if err != nil {
		return req, errors.NewCommonEdgeXWrapper(err)
	}
	return req, nil
}
```

Reply:
No, @weichou1229 ,  Lenny means users can only provide the `name` field

```
func NewAddDeviceRequest(name string) AddDeviceRequest {
	return AddDeviceRequest{
		BaseRequest: common.NewBaseRequest(),
		Device:       dtos.NewDevice(name),
	}
}

func NewDevice(name string) Device {
	return Device{
		Versionable: common.NewVersionable(),
		Id:                uuid.NewString(),
                Name:          name,
	}
}
```

Then, users still have to fill other fields before sending out.

```
 request.Device.xxx= "xxx"
 request.Device.yyy= "yyy"
```